File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.cpp
class Solution {
public:
    vector<vector<int>>dp;
    vector<vector<int>>freq;
    int mod=1e9+7;
    long recur(vector<string>&words,string target,int k,int idx){
        if(idx==target.size())return 1;
        if(k== words[0].size() || words[0].size() - k < target.size() - idx)return 0;
        if(dp[k][idx]!=-1)return dp[k][idx];
        long cnt=0;
        
        int currPos=target[idx]-'a';
        cnt=cnt+recur(words,target,k+1,idx);
        cnt += freq[k][currPos] * recur(words, target, k + 1, idx + 1);
        // cnt=(cnt+(freq[k][currPos]*(recur(words,target,k+1,idx+1))%mod)%mod);
        
        // for(int i=0;i<words.size();i++){
        //     for(int j=k;j<words[i].size();j++){
        //         if(target[idx]==words[i][j]){
        //           cnt=(cnt+recur(words,target,j+1,idx+1)%mod)%mod;
        //         }
        //     }
        // }
        return dp[k][idx]= cnt%mod;
    }
    
    int numWays(vector<string>& words, string target) {
        dp.resize(words[0].size(),vector<int>(target.size(),-1));
        freq.resize(words[0].size(),vector<int>(26,0));
        for(int i=0;i<words.size();i++){
            for(int j=0;j<words[i].size();j++){
                freq[j][words[i][j]-'a']++;
            }
        }
        return recur(words,target,0,0);
    }
};

File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.java
class Solution {
    
    
    int [][]dp;
    int [][]freq;
    int mod=1_000_000_007;
    public int recur(String[]words,String target,int k,int idx)
    {
        if(idx==target.length())return 1;
        if(k==words[0].length() || words[0].length()-k<target.length()-idx)return 0;
        if(dp[k][idx]!=-1)return dp[k][idx];
        
        int currPos=target.charAt(idx)-'a';
        
        
        long cnt=recur(words,target,k+1,idx);
        if(freq[k][currPos]>0){
            // cnt=(cnt+((recur(words,target,k+1,idx+1))%mod*freq[k][currPos])%mod)%mod;
            cnt+=(long)recur(words,target,k+1,idx+1)*freq[k][currPos];
            cnt%=mod;
        }
        // cnt=(cnt+(freq[k][currPos]*(recur(words,target,k+1,idx+1)%mod)%mod)%mod);
        
        return dp[k][idx]=(int)cnt;
      
    }
    
    public int numWays(String[] words, String target) {
        dp=new int[words[0].length()][target.length()];
        
        for(int[] i:dp){
            Arrays.fill(i,-1);
        }
        freq=new int[words[0].length()][128];
        for(int i=0;i<words.length;i++){
            for(int j=0;j<words[0].length();j++){
                freq[j][words[i].charAt(j)-'a']++;
            }
        }
       
        return recur(words,target,0,0);
        
    }
}